pylearner.ts

```diff
@@ -1350,7 +1350,119 @@ class ArrayTypeExpression extends TypeExpression {
+ * 
+ * @author Arthur Spillebeen
+ */
+class MayAliasRelation {
+  /**  Set of may-alias-sets where a may-alias-set is a collection of possible aliases. If two variables are in one of the may-alias-sets together, these variables are possible aliases.
+   *
+   * @invar a may-alias-set is never a subset of another may-aliast-set
+   * @invar a may-alias-set consists of at least two variables
+   */
+   private readonly mayAliasSets: Immutable.Set<Immutable.Set<string>>;
+
+  /**
+   * Constructor to create may-alias-relation that relates variable names if and only if variables are potential aliases by given links.
+   * 
+   * @param mayAliasSets may-alias-sets to be initialized with
+   */
+  constructor(mayAliasSets?: Immutable.Set<Immutable.Set<string>>) {
+    if (mayAliasSets) {
+      this.mayAliasSets = mayAliasSets.filter(s => s.size > 1);
+      this.mayAliasSets = this.removeSubsets();
+    } else 
+      this.mayAliasSets = ImmutableSet();
+  }
+
+  /**
+   * Returns a relation that relates variables x and y if and only if `this` holds for x and y and both x and y are not elements of `variables`.
+   * 
+   * @param variables variables to remove all links to
+   * @returns  relation with links to variables removed
+   */
+  removeVariables(variables: Immutable.Set<string>): MayAliasRelation {
+    let newMayAliasSets = this.mayAliasSets.map(s => s.subtract(s.intersect(variables)));
+    let newMayAliasRelation = new MayAliasRelation(newMayAliasSets);
+    return newMayAliasRelation;
+  }
+  
+  /**
+   * Returns a relation where a link between two variables is added to the relation. `lv` is linked to `rv`.    
+   * 
+   * @param lv lefthandside variable of the assigment of two variables
+   * @param rv rigthhandside variable of the assigment of two variables
+   * @returns new relation with given aliases linked and included
+   */
+  addMayAliasLink(lv: string, rv: string): MayAliasRelation {
+    let variables = this.mayAliasSets.flatten();
+    const rvHasAlias = variables.includes(rv);
+    let newMayAliasSets : Immutable.Set<Immutable.Set<string>> = ImmutableSet();
+    if (!rvHasAlias) {
+      let newMayAliasSet = Immutable.Set.of(lv, rv);
+      newMayAliasSets = this.mayAliasSets.add(newMayAliasSet);
+    } else if (rvHasAlias) {
+      this.mayAliasSets.forEach(s => {
+        if (s.includes(rv))
+          s = s.add(lv);
+        newMayAliasSets = newMayAliasSets.add(s);
+      });
+    }
+    const newMayAliasRelation = new MayAliasRelation(newMayAliasSets);
+    return newMayAliasRelation;
+  }
+
+  /**
+   * Returns the set of may-alias-sets where each may-alias-set is not a subset of another may-alias-set in the set, besides itself.
+   * 
+   * @returns set without subsets.
+   */
+  private removeSubsets(): Immutable.Set<Immutable.Set<string>> {
+    return this.mayAliasSets.filter(set => !this.mayAliasSets.some(otherSet => otherSet.isSuperset(set) && (set.size < otherSet.size)));
```

Comment:
```suggestion
    return this.mayAliasSets.filter(set => !this.mayAliasSets.some(otherSet => set.size < otherSet.size && otherSet.isSuperset(set)));
```
Doe eerst de `size`-check want die is veel sneller.